File: neetcode150/11_graphs/0130_surrounded_regions.cpp
/*
 * =====================================================================
 * LeetCode 130 : Surrounded Regions                                   (Medium)
 * https://leetcode.com/problems/surrounded-regions/
 * Category   : Graphs
 * ---------------------------------------------------------------------
 * PROBLEM
 *   An m×n board of 'X' and 'O'. Flip every 'O' that is NOT connected
 *   to the border through other 'O's into 'X'. Border 'O's (and any
 *   'O' reachable from them) survive.
 * ---------------------------------------------------------------------
 * INTUITION
 *   The condition "surrounded" is really about REACHABILITY: an 'O'
 *   must be flipped iff it cannot walk to the board's edge. So flip
 *   the problem — instead of testing every 'O', find all 'O's that
 *   CAN reach the border (multi-source search from every border 'O'),
 *   mark them, then flip every unmarked 'O'. That is one flood fill
 *   pass instead of a per-cell search.
 * ---------------------------------------------------------------------
 * APPROACH 1 — BFS from every border 'O' (BEST)
 *   Queue all border 'O's, mark 'E' (escape). BFS through 4-direction
 *   neighbours; any 'O' reached also escapes. Finally flip all 'O'
 *   that were never marked 'E'.
 *   Time  Complexity : O(m·n)  — every cell touched a constant times.
 *   Space Complexity : O(m·n)  — worst-case queue + board mutation.
 * ---------------------------------------------------------------------
 * APPROACH 2 — DFS recursion from every border 'O'
 *   Same reachability idea, but recursion does the traversal. The
 *   call stack can reach O(m·n) depth on a snake-shaped island —
 *   real risk of stack overflow on large boards.
 *   Time  Complexity : O(m·n)   Space Complexity : O(m·n) (stack).
 * ---------------------------------------------------------------------
 * DATA STRUCTURE DEEP DIVE — grid graphs & the dirs array
 *   A grid IS a graph: cell (r,c) is a node; edges go to its 4
 *   neighbours. The canonical trick is the DIRS array:
 *         dirs = {(-1,0),(1,0),(0,-1),(0,1)}
 *   — one loop over 4 offsets replaces four copy-pasted branches.
 *   BFS vs DFS here: identical worst-case complexity; BFS avoids the
 *   recursion depth trap of a long winding island. DFS is slightly
 *   simpler to write recursively. Both must guard bounds BEFORE
 *   probing the cell (r<0 || r>=m || c<0 || c>=n → skip).
 * =====================================================================
 */

#include <iostream>         // std::cout — printing benchmark results
#include <vector>           // std::vector — the 2D board
#include <string>           // std::string — board rows
#include <queue>            // std::queue — BFS frontier
#include <unordered_map>    // std::unordered_map — hash table (chaining)
#include <unordered_set>    // std::unordered_set — visited set
#include <chrono>           // std::chrono — high-resolution timing
#include <sys/resource.h>   // getrusage — peak RAM measurement
#include <utility>          // std::pair — (r,c) coordinates

/* =====================================================================
 * APPROACH 1 : Multi-source BFS from every border 'O'
 * ===================================================================== */
class Solution_1 {
public:
    /*
     * Purpose : Flip surrounded 'O' regions into 'X' in place.
     * Inputs  : board — m×n grid of 'X'/'O' (mutated in place).
     * Output  : void — board modified so only border-connected 'O'
     *           survive.
     */
    void solve(std::vector<std::vector<char>>& board) {
        int m = static_cast<int>(board.size());
        if (m == 0) return;              // empty board — nothing to do
        int n = static_cast<int>(board[0].size());

        // Frontier of every border 'O' — a multi-source BFS start set.
        std::queue<std::pair<int, int>> q;
        for (int r = 0; r < m; ++r) {
            for (int c = 0; c < n; ++c) {
                // Only cells ON the rim can seed the escape wave.
                bool border = (r == 0 || r == m - 1 || c == 0 || c == n - 1);
                if (border && board[r][c] == 'O') {
                    board[r][c] = 'E';   // 'E' = escapes (temporary mark)
                    q.push({r, c});
                }
            }
        }

        // dirs : the 4 cardinal offsets — one loop over 4 neighbours.
        const int dirs[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        // BFS: any 'O' reachable from a border 'O' escapes too.
        while (!q.empty()) {
            auto [r, c] = q.front();
            q.pop();
            for (int d = 0; d < 4; ++d) {
                int nr = r + dirs[d][0];
                int nc = c + dirs[d][1];
                // Bounds guard FIRST — never index outside the grid.
                if (nr < 0 || nr >= m || nc < 0 || nc >= n) continue;
                if (board[nr][nc] != 'O') continue; // only unvisited 'O'
                board[nr][nc] = 'E';     // claim: this cell escapes
                q.push({nr, nc});
            }
        }

        // Final sweep: 'E' reverts to 'O'; any remaining 'O' is truly
        // enclosed → flip to 'X'.
        for (int r = 0; r < m; ++r) {
            for (int c = 0; c < n; ++c) {
                if (board[r][c] == 'E') board[r][c] = 'O';
                else if (board[r][c] == 'O') board[r][c] = 'X';
            }
        }
    }
};

/* =====================================================================
 * APPROACH 2 : Recursive DFS from every border 'O'
 * ===================================================================== */
class Solution_2 {
private:
    // Member state so the recursive helper needs no extra parameters.
    int m_, n_;
    std::vector<std::vector<char>>* b_;

    /*
     * Purpose : Flood-fill 'O' cells reachable from (r,c) with 'E'.
     * Inputs  : r, c — current cell; preconditions: in bounds, is 'O'.
     * Output  : void — mutates the board through pointer member b_.
     */
    void dfs(int r, int c) {
        // Recursion guard: stop at walls, visited cells, or the rim.
        if (r < 0 || r >= m_ || c < 0 || c >= n_) return;
        if ((*b_)[r][c] != 'O') return;
        (*b_)[r][c] = 'E';   // mark as escaping BEFORE recursing — the
                             // classic "mark-on-entry" anti-cycle guard.
        dfs(r - 1, c);       // up
        dfs(r + 1, c);       // down
        dfs(r, c - 1);       // left
        dfs(r, c + 1);       // right
    }

public:
    /*
     * Purpose : Flip surrounded 'O' regions into 'X' in place.
     * Inputs  : board — m×n grid of 'X'/'O' (mutated in place).
     * Output  : void — board modified so only border-connected 'O'
     *           survive.
     */
    void solve(std::vector<std::vector<char>>& board) {
        m_ = static_cast<int>(board.size());
        if (m_ == 0) return;
        n_ = static_cast<int>(board[0].size());
        b_ = &board;

        // Seed DFS from every border 'O' — same multi-source idea as
        // BFS, just recursive traversal.
        for (int r = 0; r < m_; ++r) {
            for (int c = 0; c < n_; ++c) {
                bool border = (r == 0 || r == m_ - 1 || c == 0 || c == n_ - 1);
                if (border && board[r][c] == 'O') dfs(r, c);
            }
        }

        // Same final sweep: 'E' → 'O', leftover 'O' → 'X'.
        for (int r = 0; r < m_; ++r) {
            for (int c = 0; c < n_; ++c) {
                if (board[r][c] == 'E') board[r][c] = 'O';
                else if (board[r][c] == 'O') board[r][c] = 'X';
            }
        }
    }
};

/* =====================================================================
 * BENCHMARK — run both approaches on 3 tests, compare full boards
 * ===================================================================== */
int main() {
    // Test 1 (official): center 2×2 enclosed; border ring survives.
    std::vector<std::vector<char>> b1 = {
        {'X', 'X', 'X', 'X'},
        {'X', 'O', 'O', 'X'},
        {'X', 'X', 'O', 'X'},
        {'X', 'O', 'X', 'X'}
    };
    // Test 2 (official): single border 'O' → nothing flipped.
    std::vector<std::vector<char>> b2 = {
        {'X'}
    };
    // Test 3: 'O' column touching the rim saves the whole corridor.
    std::vector<std::vector<char>> b3 = {
        {'O', 'X', 'X'},
        {'O', 'X', 'O'},
        {'O', 'O', 'O'}
    };
    std::vector<std::vector<std::vector<char>>> tests = {b1, b2, b3};

    bool allPass = true;
    for (int t = 0; t < 3; ++t) {
        // --- Approach 1 timing + memory (on a fresh copy each time) ---
        auto boardA = tests[t];
        struct rusage before1, after1;
        getrusage(RUSAGE_SELF, &before1);
        auto s1 = std::chrono::high_resolution_clock::now();
        Solution_1().solve(boardA);
        auto e1_ = std::chrono::high_resolution_clock::now();
        getrusage(RUSAGE_SELF, &after1);

        // --- Approach 2 timing + memory ---
        auto boardB = tests[t];
        struct rusage before2, after2;
        getrusage(RUSAGE_SELF, &before2);
        auto s2 = std::chrono::high_resolution_clock::now();
        Solution_2().solve(boardB);
        auto e2_ = std::chrono::high_resolution_clock::now();
        getrusage(RUSAGE_SELF, &after2);

        long mem1 = (after1.ru_maxrss - before1.ru_maxrss);
        long mem2 = (after2.ru_maxrss - before2.ru_maxrss);
        double us1 = std::chrono::duration_cast<std::chrono::nanoseconds>(e1_ - s1).count() / 1000.0;
        double us2 = std::chrono::duration_cast<std::chrono::nanoseconds>(e2_ - s2).count() / 1000.0;

        // Both approaches must produce IDENTICAL final boards.
        bool same = (boardA == boardB);
        // Spot-verify against the known official result for test 1.
        bool known = true;
        if (t == 0) {
            std::vector<std::vector<char>> expect = {
                {'X', 'X', 'X', 'X'},
                {'X', 'X', 'X', 'X'},
                {'X', 'X', 'X', 'X'},
                {'X', 'O', 'X', 'X'}
            };
            known = (boardA == expect);
        }
        bool ok = same && known;
        allPass = allPass && ok;

        std::cout << "Test " << (t + 1) << (ok ? " PASS" : " FAIL")
                  << " | A1 time=" << us1 << " µs mem=" << mem1 << " KB"
                  << " | A2 time=" << us2 << " µs mem=" << mem2 << " KB\n";
        if (!ok) {
            std::cout << "  A1 board:\n";
            for (const auto& row : boardA) {
                for (char c : row) std::cout << c;
                std::cout << "\n";
            }
            std::cout << "  A2 board:\n";
            for (const auto& row : boardB) {
                for (char c : row) std::cout << c;
                std::cout << "\n";
            }
        }
    }
    std::cout << (allPass ? "PASS : all tests green on both approaches.\n"
                          : "FAIL : at least one test/approach disagrees.\n");
    return 0;
}

File: neetcode150/11_graphs/0130_surrounded_regions.py
"""
=====================================================================
LeetCode 130 : Surrounded Regions                               (Medium)
https://leetcode.com/problems/surrounded-regions/
Category   : Graphs
---------------------------------------------------------------------
PROBLEM
    An m×n board of 'X' and 'O'. Flip every 'O' that is NOT connected
    to the border through other 'O's into 'X'. Border 'O's (and any
    'O' reachable from them) survive.
---------------------------------------------------------------------
INTUITION
    "Surrounded" is really REACHABILITY: an 'O' must flip iff it
    cannot walk to the board's edge. So invert the question — multi-
    source flood fill from every border 'O', mark those cells as
    escaping, then flip every unmarked 'O'. One pass, no per-cell
    searches.
---------------------------------------------------------------------
APPROACH 1 — BFS from every border 'O' (BEST)
    Queue all border 'O's, mark 'E'. BFS the 4 neighbours; every 'O'
    reached escapes too. Final sweep: 'E'→'O', leftover 'O'→'X'.
    Time  Complexity : O(m·n)   Space Complexity : O(m·n)
---------------------------------------------------------------------
APPROACH 2 — Recursive DFS from every border 'O'
    Identical idea, recursive traversal. Call stack can reach O(m·n)
    on a winding island — stack-overflow risk on big boards.
    Time  Complexity : O(m·n)   Space Complexity : O(m·n) (stack)
---------------------------------------------------------------------
DATA STRUCTURE DEEP DIVE — grid graphs & the dirs array
    A grid IS a graph: cell (r,c) is a node, edges to its 4
    neighbours. The dirs array `[(−1,0),(1,0),(0,−1),(0,1)]` turns 4
    copy-pasted branches into one loop. Bounds must be checked BEFORE
    indexing the grid. BFS vs DFS: identical O(m·n); BFS avoids deep
    recursion; DFS is terser to write. Both need a visited marker to
    stop cycles (here: repainting 'O'→'E').
=====================================================================
REAL LIFE
    • Flood-fill tools (paint bucket in every editor) are BFS/DFS on
      a grid graph.
    • "Is this pixel/region reachable from the edge?" = leak
      detection in circuit-board etching and PCB design.
    • Go/Baduk territory evaluation: stones enclosed by the opponent.
=====================================================================
"""

from __future__ import annotations
import time
import tracemalloc
from collections import deque
from typing import List

# =====================================================================
# APPROACH 1 : Multi-source BFS from every border 'O'
# =====================================================================
class SolutionOne:
    """
    Purpose : Flip surrounded 'O' regions into 'X' in place.
    Inputs  : board — m×n grid of 'X'/'O' (mutated in place).
    Output  : None — board modified so only border-connected 'O'
              survive.
    """

    def solve(self, board: List[List[str]]) -> None:
        m, n = len(board), len(board[0])
        # Frontier of every border 'O' — multi-source BFS start set.
        queue = deque()
        for r in range(m):
            for c in range(n):
                # Only cells ON the rim can seed the escape wave.
                if (r in (0, m - 1) or c in (0, n - 1)) and board[r][c] == "O":
                    board[r][c] = "E"      # 'E' = escapes (temporary mark)
                    queue.append((r, c))

        # dirs : the 4 cardinal offsets — one loop over 4 neighbours.
        dirs = [(-1, 0), (1, 0), (0, -1), (0, 1)]

        # BFS: any 'O' reachable from a border 'O' escapes too.
        while queue:
            r, c = queue.popleft()
            for dr, dc in dirs:
                nr, nc = r + dr, c + dc
                # Bounds guard FIRST — never index outside the grid.
                if 0 <= nr < m and 0 <= nc < n and board[nr][nc] == "O":
                    board[nr][nc] = "E"    # claim: this cell escapes
                    queue.append((nr, nc))

        # Final sweep: 'E' reverts to 'O'; any remaining 'O' is truly
        # enclosed → flip to 'X'.
        for r in range(m):
            for c in range(n):
                if board[r][c] == "E":
                    board[r][c] = "O"
                elif board[r][c] == "O":
                    board[r][c] = "X"

# =====================================================================
# APPROACH 2 : Recursive DFS from every border 'O'
# =====================================================================
class SolutionTwo:
    """
    Purpose : Flip surrounded 'O' regions into 'X' in place.
    Inputs  : board — m×n grid of 'X'/'O' (mutated in place).
    Output  : None — board modified so only border-connected 'O'
              survive.
    """

    def solve(self, board: List[List[str]]) -> None:
        m, n = len(board), len(board[0])

        # Nested closure keeps the recursion signature tiny — only the
        # coordinates travel down the call stack.
        def dfs(r: int, c: int) -> None:
            # Recursion guard: stop at walls, visited cells, or rim.
            if r < 0 or r >= m or c < 0 or c >= n or board[r][c] != "O":
                return
            board[r][c] = "E"              # mark BEFORE recursing — the
                                           # classic anti-cycle guard.
            dfs(r - 1, c)                  # up
            dfs(r + 1, c)                  # down
            dfs(r, c - 1)                  # left
            dfs(r, c + 1)                  # right

        # Seed DFS from every border 'O' — same multi-source idea.
        for r in range(m):
            for c in range(n):
                if (r in (0, m - 1) or c in (0, n - 1)) and board[r][c] == "O":
                    dfs(r, c)

        # Same final sweep: 'E' → 'O', leftover 'O' → 'X'.
        for r in range(m):
            for c in range(n):
                if board[r][c] == "E":
                    board[r][c] = "O"
                elif board[r][c] == "O":
                    board[r][c] = "X"

# =====================================================================
# BENCHMARK — run both approaches on 3 tests, compare full boards
# =====================================================================
if __name__ == "__main__":
    # Test 1 (official): center 2×2 enclosed; border ring survives.
    t1 = [["X", "X", "X", "X"],
          ["X", "O", "O", "X"],
          ["X", "X", "O", "X"],
          ["X", "O", "X", "X"]]
    # Test 2 (official): single border 'O' → nothing flipped.
    t2 = [["X"]]
    # Test 3: 'O' column touching the rim saves the whole corridor.
    t3 = [["O", "X", "X"],
          ["O", "X", "O"],
          ["O", "O", "O"]]
    tests = [t1, t2, t3]
    # Official expected board for test 1.
    expected_t1 = [["X", "X", "X", "X"],
                   ["X", "X", "X", "X"],
                   ["X", "X", "X", "X"],
                   ["X", "O", "X", "X"]]

    all_pass = True
    for idx, test in enumerate(tests, start=1):
        # --- Approach 1: time + peak memory (fresh copy each run) ---
        board_a = [row[:] for row in test]
        tracemalloc.start()
        t0 = time.perf_counter()
        SolutionOne().solve(board_a)
        t1_ = time.perf_counter()
        _, peak1 = tracemalloc.get_traced_memory()
        tracemalloc.stop()

        # --- Approach 2: time + peak memory ---
        board_b = [row[:] for row in test]
        tracemalloc.start()
        t2_ = time.perf_counter()
        SolutionTwo().solve(board_b)
        t3_ = time.perf_counter()
        _, peak2 = tracemalloc.get_traced_memory()
        tracemalloc.stop()

        # Both approaches must produce IDENTICAL final boards, and test
        # 1 must match the official expected result.
        same = board_a == board_b
        known = board_a == expected_t1 if idx == 1 else True
        ok = same and known
        all_pass = all_pass and ok
        print(
            f"Test {idx} {'PASS' if ok else 'FAIL'} "
            f"| A1 time={(t1_-t0)*1e6:.2f} µs peak={peak1/1024:.1f} KB"
            f" | A2 time={(t3_-t2_)*1e6:.2f} µs peak={peak2/1024:.1f} KB"
            f" | boards identical: {same}"
        )

    print("PASS : all tests green on both approaches."
          if all_pass else "FAIL : at least one test/approach disagrees.")